File: database/models/PlayerGameStats.py
"""Defines the comments object and provides functions to get and manipulate one"""
import time

from database import db
from database.models import Games

# create table main.playerGameStats
# (
#     id                INTEGER primary key autoincrement,
# gameId            INTEGER references main.games,
# playerId          INTEGER references main.people,
# teamId            INTEGER references main.teams,
# opponentId        INTEGER references main.teams,
# tournamentId      INTEGER references main.tournaments,
# roundsPlayed      INTEGER,
# roundsBenched     INTEGER,
# isBestPlayer      INTEGER,
# isFinal           INTEGER,
# points            INTEGER,
# aces              INTEGER,
# faults            INTEGER,
# servedPoints      INTEGER,
# servedPointsWon   INTEGER,
# servesReceived    INTEGER,
# servesReturned    INTEGER,
# doubleFaults      INTEGER,
# greenCards        INTEGER,
# warnings          INTEGER,
# yellowCards       INTEGER,
# redCards          INTEGER,
# cardTimeRemaining INTEGER,
# cardTime          INTEGER,
# startSide         TEXT
# );

PERCENTAGES = [
    "Return Rate"
]


class PlayerGameStats(db.Model):
    __tablename__ = "playerGameStats"

    # Auto-initialised fields
    id = db.Column(db.Integer(), primary_key=True)
    created_at = db.Column(db.Integer(), default=time.time, nullable=False)

    # Set fields
    game_id = db.Column(db.Integer(), db.ForeignKey("games.id"), nullable=False)
    player_id = db.Column(db.Integer(), db.ForeignKey("people.id"), nullable=False)
    team_id = db.Column(db.Integer(), db.ForeignKey("teams.id"), nullable=False)
    opponent_id = db.Column(db.Integer(), db.ForeignKey("teams.id"), nullable=False)
    tournament_id = db.Column(db.Integer(), db.ForeignKey("tournaments.id"), nullable=False)
    rounds_on_court = db.Column(db.Integer(), default=0, nullable=False)
    rounds_carded = db.Column(db.Integer(), default=0, nullable=False)
    points_scored = db.Column(db.Integer(), default=0, nullable=False)
    aces_scored = db.Column(db.Integer(), default=0, nullable=False)
    is_best_player = db.Column(db.Boolean(), default=0, nullable=False)
    faults = db.Column(db.Integer(), default=0, nullable=False)
    double_faults = db.Column(db.Integer(), default=0, nullable=False)
    served_points = db.Column(db.Integer(), default=0, nullable=False)
    served_points_won = db.Column(db.Integer(), default=0, nullable=False)
    serves_received = db.Column(db.Integer(), default=0, nullable=False)
    serves_returned = db.Column(db.Integer(), default=0, nullable=False)
    ace_streak = db.Column(db.Integer(), default=0, nullable=False)
    serve_streak = db.Column(db.Integer(), default=0, nullable=False)
    warnings = db.Column(db.Integer(), default=0, nullable=False)
    green_cards = db.Column(db.Integer(), default=0, nullable=False)
    yellow_cards = db.Column(db.Integer(), default=0, nullable=False)
    red_cards = db.Column(db.Integer(), default=0, nullable=False)
    card_time = db.Column(db.Integer(), default=0, nullable=False)
    card_time_remaining = db.Column(db.Integer(), default=0, nullable=False)
    rating = db.Column(db.Integer(), default=3)
    start_side = db.Column(db.Text())
    side_of_court = db.Column(db.Text())

    tournament = db.relationship("Tournaments", foreign_keys=[tournament_id])
    player = db.relationship("People", foreign_keys=[player_id])
    team = db.relationship("Teams", foreign_keys=[team_id])
    opponent = db.relationship("Teams", foreign_keys=[opponent_id])
    game = db.relationship("Games", foreign_keys=[game_id])

    rows = {
        "Rounds on Court": rounds_on_court,
        "Rounds Carded": rounds_carded,
        "Points Scored": points_scored,
        "Aces Scored": aces_scored,
        "Faults": faults,
        "Double Faults": double_faults,
        "Served Points": served_points,
        "Served Points Won": served_points_won,
        "Serves Received": serves_received,
        "Serves Returned": serves_returned,
        "Warnings": warnings,
        "Green Cards": green_cards,
        "Yellow Cards": yellow_cards,
        "Red Cards": red_cards,
        "Cards": green_cards + yellow_cards + red_cards,
        "Elo": None,
        "Elo Delta": None,
        "Result": Games.winning_team_id == team_id,
        "IGA Side": Games.iga_side_id == team_id,
        "Ranked": Games.ranked,
        "Return Rate": serves_returned / serves_received
    }

    def reset_stats(self):
        self.rounds_on_court = 0
        self.rounds_carded = 0
        self.points_scored = 0
        self.aces_scored = 0
        self.faults = 0
        self.double_faults = 0
        self.served_points = 0
        self.served_points_won = 0
        self.serves_received = 0
        self.serves_returned = 0
        self.warnings = 0
        self.green_cards = 0
        self.yellow_cards = 0
        self.red_cards = 0
        self.card_time = 0
        self.card_time_remaining = 0
        self.is_best_player = 0

    def stats(self, admin=False):
        from database.models.GameEvents import GameEvents
        from database.models.EloChange import EloChange
        first_ge = GameEvents.query.filter(GameEvents.game_id == self.game_id).first()
        elo_delta = EloChange.query(EloChange.game_id == self.game_id, EloChange.player_id == self.player_id).first()
        d = self.game.stats() | {
            "Rounds on Court": self.rounds_on_court,
            "Ranked": self.game.ranked,
            "Rounds Carded": self.rounds_carded,
            "Points Scored": self.points_scored,
            "Aces Scored": self.aces_scored,
            "Faults": self.faults,
            "Double Faults": self.double_faults,
            "Served Points": self.served_points,
            "Served Points Won": self.served_points_won,
            "Serves Received": self.serves_received,
            "Serves Returned": self.serves_returned,
            "Warnings": self.warnings,
            "Green Cards": self.green_cards,
            "Yellow Cards": self.yellow_cards,
            "Red Cards": self.red_cards,
            "Cards": self.red_cards + self.yellow_cards + self.green_cards,
            "Elo": round(self.player.elo(self.game_id), 2),
            "Elo Delta": round(elo_delta.elo_delta, 2),
            "Result": int(self.team_id == self.game.winning_team_id),
            "IGA Side": int(self.team_id == self.game.iga_side_id),
            "Served First": first_ge and int(self.team_id == first_ge.team_to_serve_id),
            "Return Rate": self.serves_returned / (self.serves_received or 1),
            "Timeouts Used": (
                self.game.team_one_timeouts if self.game.team_one_id == self.team_id else self.game.team_two_timeouts),
        }
        if admin:
            d["Penalty Points"] = self.green_cards * 2 + self.yellow_cards * 5 + self.red_cards * 10
            d["Rating"] = self.rating
        return d

    @classmethod
    def row_by_name(cls, name):
        return cls.rows[name]

    def as_dict(self, include_game=True, make_nice=False, admin_view=False, official_view=False, include_stats=False,
                include_prev_cards=False):
        from database.models.EloChange import EloChange
        from database.models.GameEvents import GameEvents
        elo_delta = EloChange.query.filter(EloChange.game_id == self.game_id,
                                           EloChange.player_id == self.player_id).first()
        d = {
            "isBestPlayer": bool(self.is_best_player),
            "cardTime": self.card_time,
            "cardTimeRemaining": self.card_time_remaining,
            "sideOfCourt": self.side_of_court,
            "isCaptain": self.team.captain_id == self.player_id,
            "startSide": self.start_side,
        }
        if admin_view:
            d["rating"] = self.rating or 3
        if include_prev_cards and (admin_view or official_view):
            cards: list[GameEvents] = GameEvents.query.filter(GameEvents.tournament_id == self.tournament_id,
                                                              GameEvents.player_id == self.player_id, GameEvents.is_card == True).all()
            d["prevCards"] = [i.as_dict(include_game=False, include_player=False, card_details=True) for i in cards]
        if include_stats:
            stats = {
                "Rounds on Court": self.rounds_on_court,
                "Rounds Carded": self.rounds_carded,
                "Points Scored": self.points_scored,
                "Aces Scored": self.aces_scored,
                "Faults": self.faults,
                "Double Faults": self.double_faults,
                "Served Points": self.served_points,
                "Served Points Won": self.served_points_won,
                "Serves Received": self.serves_received,
                "Serves Returned": self.serves_returned,
                "Biggest Ace Streak": self.ace_streak,
                "Biggest Serve Streak": self.serve_streak,
                "Green Cards": self.green_cards,
                "Yellow Cards": self.yellow_cards,
                "Red Cards": self.red_cards,
                "Starting Side": self.start_side,
                "Elo": round(self.player.elo(self.game_id), 2),
                "Elo Delta": round(elo_delta.elo_delta if elo_delta else 0, 2),
            }
            if make_nice:
                for k, v in stats.items():
                    if k in PERCENTAGES:
                        d[k] = f"{100.0 * v: .2f}%".strip()
                    elif isinstance(v, float):
                        d[k] = round(v, 2)
            d["stats"] = stats

        if include_game:
            d["team"] = self.team.as_dict(),
            d["game"] = self.game.as_dict()

        return d | self.player.as_dict()

File: HandballBackend/Database/Models/PlayerGameStats.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using HandballBackend.Database.SendableTypes;
using HandballBackend.Models;

namespace HandballBackend.Database.Models;

[Table("playerGameStats", Schema = "main")]
public class PlayerGameStats {
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    [Column("id")]
    public int Id { get; set; }

    [Required]
    [Column("game_id")]
    public int GameId { get; set; }

    [Required]
    [Column("player_id")]
    public int PlayerId { get; set; }

    [Required]
    [Column("team_id")]
    public int TeamId { get; set; }

    [Required]
    [Column("opponent_id")]
    public int? OpponentId { get; set; }

    [Required]
    [Column("tournament_id")]
    public int TournamentId { get; set; }

    [Required]
    [Column("rounds_on_court")]
    public int RoundsOnCourt { get; set; } = 0;

    [Required]
    [Column("rounds_carded")]
    public int RoundsCarded { get; set; } = 0;

    [Required]
    [Column("points_scored")]
    public int PointsScored { get; set; } = 0;

    [Required]
    [Column("aces_scored")]
    public int AcesScored { get; set; } = 0;

    [Required]
    [Column("faults")]
    public int Faults { get; set; } = 0;

    [Required]
    [Column("served_points")]
    public int ServedPoints { get; set; } = 0;

    [Required]
    [Column("served_points_won")]
    public int ServedPointsWon { get; set; } = 0;

    [Required]
    [Column("serves_received")]
    public int ServesReceived { get; set; } = 0;

    [Required]
    [Column("serves_returned")]
    public int ServesReturned { get; set; } = 0;

    [Required]
    [Column("double_faults")]
    public int DoubleFaults { get; set; } = 0;

    [Required]
    [Column("green_cards")]
    public int GreenCards { get; set; } = 0;

    [Required]
    [Column("warnings")]
    public int Warnings { get; set; } = 0;

    [Required]
    [Column("yellow_cards")]
    public int YellowCards { get; set; } = 0;

    [Required]
    [Column("red_cards")]
    public int RedCards { get; set; } = 0;

    [Required]
    [Column("card_time_remaining")]
    public int CardTimeRemaining { get; set; } = 0;

    [Required]
    [Column("card_time")]
    public int CardTime { get; set; } = 0;

    [Column("start_side", TypeName = "TEXT")]
    public string StartSide { get; set; }

    [Required]
    [Column("created_at")]
    public int CreatedAt { get; set; } = (int) DateTimeOffset.UtcNow.ToUnixTimeSeconds();

    [Required]
    [Column("is_best_player")]
    public bool IsBestPlayer { get; set; } = false;

    [Required]
    [Column("ace_streak")]
    public int AceStreak { get; set; } = 0;

    [Required]
    [Column("serve_streak")]
    public int ServeStreak { get; set; } = 0;

    [Column("side_of_court")]
    public string? SideOfCourt { get; set; }

    [Column("rating")]
    public int? Rating { get; set; }

    [ForeignKey("GameId")]
    public Game Game { get; set; }

    [ForeignKey("PlayerId")]
    public Person Player { get; set; }

    [ForeignKey("TeamId")]
    public Team Team { get; set; }

    [ForeignKey("OpponentId")]
    public Team Opponent { get; set; }

    [ForeignKey("TournamentId")]
    public Tournament Tournament { get; set; }

    public GamePlayerData ToSendableData(bool includeStats = false, bool formatData = false) {
        return new GamePlayerData(this, includeStats, formatData);
    }

    public void ResetStats() {
        RoundsOnCourt = 0;
        RoundsCarded = 0;
        PointsScored = 0;
        AcesScored = 0;
        Faults = 0;
        DoubleFaults = 0;
        ServedPoints = 0;
        ServedPointsWon = 0;
        ServesReceived = 0;
        ServesReturned = 0;
        Warnings = 0;
        GreenCards = 0;
        YellowCards = 0;
        RedCards = 0;
        CardTime = 0;
        CardTimeRemaining = 0;
        IsBestPlayer = false;
    }
}